### ISSUE TITLE
refactor(tune): validated `TrainCtx` — replace ten-argument train-tape signatures

### ISSUE THREAD

**Issue by ohdearquant** [ohdearquant/lattice#844]
> _Issue authored by Claude (Anthropic agent) on behalf of @ohdearquant._

## What

Introduce a validated, immutable run context in `crates/tune/src/lora/train_core.rs`
that replaces the wide positional-argument signatures on the training tape's core
entry points, and remove the `#[allow(clippy::too_many_arguments)]` markers those
signatures currently require.

On current `main`, the merged GDN-LoRA gradient work made every core entry point
wider:

- `forward_full` takes 10 positional arguments (`ctx, layers, loras, gdn_loras, head,
  d, gdn_dims, cfg, rank, scale`).
- `eval_chain_nll` takes the same 10.
- `nll_and_grads` takes 10 (`fwd, layers, loras, head, d, gdn_dims, num_slots,
  num_gdn_slots, rank, scale`) and carries `#[allow(clippy::too_many_arguments)]`.
- `apply_adam_updates` (8 args) and `apply_gdn_adam_updates` (8 args, also
  `#[allow(clippy::too_many_arguments)]`) are separate helpers for the GQA and GDN
  LoRA slot families with no shared context object.

Replace this with a nested, validated context:

```rust
pub struct TrainCtx<'a> {
    geometry: TapeGeometry<'a>,
    lora: LoraExecution,
    slots: SlotLayout<'a>,
    adam: AdamConfig,
}

pub struct TapeGeometry<'a> {
    dims: &'a Dims,
    gdn_dims: &'a GdnDims,
    model: &'a Qwen35Config,
}

struct LoraExecution {
    rank: usize,
    scale: f32,
}

pub struct SlotLayout<'a> {
    gqa_layers: &'a [usize],
    gdn_layers: &'a [usize],
}

pub struct AdamConfig {
    learning_rate: f32,
    beta1: f32,
    beta2: f32,
    epsilon: f32,
}
```

`TrainCtx::try_new` validates rank/non-finite hyperparameters and geometry, derives
the private execution-only `{rank, scale}` pair from `alpha / rank`, and validates
unique/in-range slot layers plus mixer-kind/slot-index agreement. `num_gqa_slots()`
and `num_gdn_slots()` derive from the layout. `LoraExecution` is private, not
serialized or exported, and is not adapter governance (see Non-goals). Post-refactor
signatures become:

```rust
forward_full(seq, layers, gqa_loras, gdn_loras, head, train)
eval_chain_nll(seqs, layers, gqa_loras, gdn_loras, head, train)
nll_and_grads(fwd, layers, gqa_loras, head, train)
apply_adam_updates(adam, gqa_loras, grads, train)
apply_gdn_adam_updates(adam, gdn_loras, grads, train)
```

## Why

Ten-argument functions with two `#[allow(clippy::too_many_arguments)]` escapes are
themselves the defect: every future caller must reconstruct the same positional
tuple correctly by hand, and geometry/slot/optimizer values that must agree with
each other (rank vs. array shapes, slot index vs. mixer kind, GQA vs. GDN slot
counts) currently have no single place that checks that agreement. A validated
constructor turns "the caller got the tuple order right" into "the type does not
exist unless the invariants hold."

## Non-goals (explicit fences)

- No math, CLI, initialization, or serialization change. This is a signature/type
  refactor only — same forward tape, same optimizer math, same output values.
- The context may not own weights, caches, gradients, LoRA vectors, or optimizer
  state — geometry, layout, and policy only. If most functions in a follow-up only
  need one nested component, pass that component instead of the whole context.
- Do not recreate or duplicate the shared adapter descriptor scoped to issue #615
  (rank/alpha/target-module/blend governance for the tune/Metal-inference boundary).
  `TrainCtx` may consume that descriptor once #615 lands; it must not become a
  second copy of it in the meantime.

## Checklist

- [ ] Add `TrainCtx`, `TapeGeometry`, `LoraExecution`, `SlotLayout`, `AdamConfig` to
      `crates/tune/src/lora/train_core.rs` with `TrainCtx::try_new` validation
      (rank/hyperparameter finiteness, geometry consistency, unique/in-range slot
      layers, mixer-kind/slot-index agreement).
- [ ] Migrate `forward_full`, `eval_chain_nll`, `nll_and_grads`, `apply_adam_updates`,
      `apply_gdn_adam_updates` to the narrowed signatures shown above.
- [ ] Migrate call sites in `crates/tune/src/lora/train.rs` and the training bins
      (`train_grad_full.rs`; `train_grad.rs` only if it shares the affected helpers).
- [ ] Remove both `#[allow(clippy::too_many_arguments)]` markers now covered by the
      new signatures.
- [ ] Add unit tests for `TrainCtx::try_new` invariant rejection: out-of-range slot
      layer, duplicate slot layer, mixer-kind/slot-index mismatch, non-finite
      rank/alpha/learning-rate/beta/epsilon inputs.

## Gates

- Existing primitive gradchecks and the post-merge per-array GDN LoRA gradcheck
  (`overall_max_rel < 1e-2` across all ten GDN LoRA arrays) must continue to pass
  unmodified in value, only migrated to the new call shape.
- Full-tape `--gradcheck` at first layers 19 and 23, where a checkpoint runner is
  available.
- Real-model TBV (cached/assembled NLL vs. `Qwen35Model::compute_token_nlls`) must
  still pass at the new call sites.
- New `TrainCtx` invariant unit tests (above) must all fail before the fix and pass
  after — mutation-sensitive by construction, since the code changes from "this
  compiles" to "this constructor rejects the input."
- Mutation gate: swap one GQA/GDN slot index or one derived-scale use at a call
  site and confirm the targeted central-difference test fails (`max_rel` gate
  trips, or `TrainCtx::try_new` rejects it if the swap makes layout inconsistent).
- `cargo clippy --workspace -- -D warnings` clean with both `too_many_arguments`
  allows removed.

## Sequencing

First implementation PR in this series (no other unit in this series depends on
work landing before it). Does not touch `crates/tune/src/bin/train_grad_layer23.rs`
or the harness/parser consolidation — those are separate units in this series.

**Comment by ohdearquant** [comment]
Closed by PR #858 (merged): validated TrainCtx/TapeGeometry/SlotLayout/AdamConfig landed in train_core.rs with mixer-kind + non-finite-geometry validation. The PR body referenced this issue without a closing keyword, so the auto-close never fired.
